app/invocation/execution_status.tsx

```diff
@@ -0,0 +1,104 @@
+import rpc_service, { ServerStreamHandler, ServerStream } from "../service/rpc_service";
+import { execution_stats } from "../../proto/execution_stats_ts_proto";
+import { build } from "../../proto/remote_execution_ts_proto";
+
+const ExecutionStage = build.bazel.remote.execution.v2.ExecutionStage.Value;
+const ExecutionState = build.bazel.remote.execution.v2.ExecutionProgress.ExecutionState;
+
+/**
+ * Live-streams execution updates using the WaitExecution API.
+ * Each operation is unpacked into a more usable format.
+ */
+export function waitExecution(
+  executionId: string,
+  handler: ServerStreamHandler<ExecuteOperation>
+): ServerStream<ExecuteOperation> {
+  return rpc_service.service.waitExecution(
+    { executionId },
+    {
+      next: (response) => {
+        const unpackedOperation = unpackResponse(response);
+        if (unpackedOperation) {
+          handler.next(unpackedOperation);
+        }
+      },
+      error: (e) => {
+        // TODO: if we disconnected from the app (e.g. app restart), don't
+        // bubble up the error, just transparently reconnect.
+        handler.error(e);
+      },
+      complete: () => {
+        handler.complete();
+      },
+    }
+  );
+}
+
+export function executionStatusLabel(op: ExecuteOperation) {
+  if (op.done) {
+    return "Completed";
+  }
+
+  switch (op.metadata?.stage) {
+    case ExecutionStage.CACHE_CHECK:
+      return "Checking cache";
+    case ExecutionStage.QUEUED:
+      return "Queued";
+    case ExecutionStage.EXECUTING:
+      // Return fine-grained progress using BB-specific enum
```

Comment:
nit: for readability, might be better to just break out the `executionState`-based text into its own function and call it directly here--the fall-through is a bit awkward.